### PR TITLE
[herzborg] Herzborg binding

### PR DESCRIPTION
**Pull request by Sonic-Amiga**
Supports Herzborg curtain motors over RS-485 network.
Initial contribution.

Signed-off-by: Pavel Fedin <pavel_fedin@mail.ru>

### CODE REVIEW

**Review by Hilbrand — CHANGES_REQUESTED** [review]
Thank you for your contribution. I've added a few remarks. One general comment the code in the `BusHandler` seems to something that should be in a separate class not in a inherited class. That makes the code much more complexer. But that is something you might consider some other time.

File: bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/CurtainHandler.java
Comment (Sonic-Amiga) [reply bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/CurtainHandler.java:207]: Thanks, done

File: bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/HerzborgHandlerFactory.java
Comment (Sonic-Amiga) [reply bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/HerzborgHandlerFactory.java:46]: Thank you for the suggestion, these forced @Nullable's are a huge source of headace

File: bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/SerialBusHandler.java
Comment (Sonic-Amiga) [reply bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/SerialBusHandler.java:76]: Done. However, how are communication timeouts handled if enableReceiveTimeout() is not implemented? What if there's some glitch on the serial line and one byte gets swallowed by noise? This would hand the whole binding because it would keep waiting for the missing byte forever.

File: bundles/org.openhab.binding.herzborg/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Sonic-Amiga) [reply bundles/org.openhab.binding.herzborg/src/main/resources/OH-INF/thing/thing-types.xml:7]: Hopefully fixed all of them

File: bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/SerialBusHandler.java
Comment (Sonic-Amiga) [reply bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/SerialBusHandler.java:148]: I've tested on my operational 2.5.9 installation; yes, it seems to be fine, i get a couple of events per one poll (once per second), not a big deal. Had to leave an empty function definition though, it's required by the interface.

**Comment by Sonic-Amiga** [comment]
> Thank you for your contribution. I've added a few remarks. One general comment the code in the `BusHandler` seems to something that should be in a separate class not in a inherited class. That makes the code much more complexer.

 Are you sure? A Bus of any kind would always be a Bridge. In fact i had something like serial over in mind IP when i designed this code. I didn't know that it's supported by the opehab's serial abstraction layer back then.
 Anyways, this design is pretty much a placeholder for any potential expansion in some future (though nobody seems to know this vendor). In fact they are also selling a wifi dongle, which plugs into a motor and makes it possible to command it from a smartphone. I guess that thing could be supported as well, i don't know if it's a simple serlal-to-IP bridge or something more complex. Personally i currently have no plans for that, the add-on is rather costly and i have enough wires and serial ports, so i am happy with the wired connection.

**Comment by Sonic-Amiga** [comment]
@Hilbrand ping

**Comment by Hilbrand** [comment]
> A Bus of any kind would always be a Bridge

That is not the point. I'm referring to a class should not do to many different things. The class extends the `BaseBridgeHandler`. This implies that the code in that class is an extension of methods related to channel states and thing status. However all code in the class `BusHandler` is related to the communication with the device. So this code should be in a class that doesn't extend `BaseBridgeHandler` and the class should be instantiated in a subclass of `BaseBridgeHandler` or somewhere else. You are actually limiting future support because because now it's part of the handler and implies the communication is always via `Packet` data. If another mechanism would be used you can't make it flexible for the thing handlers because it can only extend 1 class. Or you would need to add all different communication ways to the class.
To give another point. You synchronize the method `doPacket` probably because this method should not be called at the same time. However because it extends `BaseBridgeHandler` it also synchronizes with all synchronized methods in `BaseBridgeHandler` and blocks any methods in that class. This can lead to unexpected results if your reading takes some time. If the code would be in a separate none extending class the synchronization would only be for the `doPacket` method.

**Comment by Sonic-Amiga** [comment]
Well, OK, you convinced me, i'll do it

**Comment by Sonic-Amiga** [comment]
@Hilbrand Well, i've done it. Frankly speaking, it does not simplify anything. Instead of one class, we now have two tightly coupled classes. Well, let's say you just love it this way.

**Comment by Sonic-Amiga** [comment]
@Hilbrand Hello! 26 days have passed, no feedback from you

**Comment by Sonic-Amiga** [comment]
Fixed spotless (yet again)

**Comment by Sonic-Amiga** [comment]
Dear @Hilbrand ! Did i offend you somehow? Sorry if so. 1.5 more months have passed, no response from you.
Anyone else? @fwolter ?

**Comment by Sonic-Amiga** [comment]
Rebased. Anyone ?

**Comment by Sonic-Amiga** [comment]
YARRR
Yet Another Rebase...

File: bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/CurtainHandler.java
Comment (Skinah) [inline bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/CurtainHandler.java:186]:
```suggestion
                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Invalid response received:" + DatatypeConverter.printHexBinary(reply.getBuffer()));
```

When the thing status is changed, it already gets logged to the event.log and it is more visible in the main UI. To prevent a double up of logging, this way is preferred.

File: bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/SerialBus.java
Comment (Skinah) [inline bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/SerialBus.java:112]: Can this be removed if the thing comes online right away it will go into the event.log automatically?

File: bundles/org.openhab.binding.herzborg/src/main/resources/OH-INF/thing/thing-types.xml
Comment (Skinah) [inline bundles/org.openhab.binding.herzborg/src/main/resources/OH-INF/thing/thing-types.xml:31]: See https://www.openhab.org/docs/developer/guidelines.html#java-coding-style

Channels should be lowerCamelCase to make all bindings consistent and make life easier for textual users.

**Comment by Skinah** [comment]
OK that seems to be everything and I have looked over all the code and it looks good to me. I'll approve once these points are addressed unless @Hilbrand has any comments? The only other comment is that the zigbee binding which is in its own repo has the ability to parse and display any detected serial ports to make it easier for the end user. Never looked at how it is done or how hard to add, but it would be cool to add that. Not necessary for me to approve, just hinting it would be nice to have.

As for how long this took, I can only say sorry and that once it is merged you also become a maintainer and I hope you will assist in doing reviews as we clearly need more people doing it. Your coding is better than mine so I hope you contribute more bindings and help out with reviews, both would be great and thanks for this contribution.

**Comment by Sonic-Amiga** [comment]
Build error again:
```
[ERROR] Failed to execute goal on project org.openhab.addons.reactor.bundles: Could not resolve dependencies for project org.openhab.addons.bundles:org.openhab.addons.reactor.bundles:pom:3.2.0-SNAPSHOT: Could not transfer artifact org.openhab.core.bundles:org.openhab.core.persistence:jar:3.2.0-20211013.065607-29 from/to openhab-snapshot (https://openhab.jfrog.io/openhab/libs-snapshot): /home/jenkins/jenkins-agent1/maven-repositories/3/org/openhab/core/bundles/org.openhab.core.persistence/3.2.0-SNAPSHOT/org.openhab.core.persistence-3.2.0-20211013.065607-29.jar.part (No such file or directory) -> [Help 1]
```
some internal dependency resolution error?

**Comment by Sonic-Amiga** [comment]
@Skinah Hello! It's been more than a year in total.
IIRC next year you'll switch copyright messages and tests will stop passing. I'll have to update the code again. Is there any sense for me to spend time on trying to contribute this code ?
 I understand that you are the Masters here and you decide, what comes in, and what doesn't. I also understand that the hardware, supported by this binding, is probably local to russian market; and you might be not interested in supporting such things.
 Please reply anything. This silence is very baffling...

**Comment by Skinah** [comment]
@fwolter Is this ok to merge once the copyright is updated?

@Sonic-Amiga 
Sorry I do not have merging rights so it needs someone to see your thread and to make the call. My personal opinion is that if someone writes and binding and will maintain it then it should get merged if someone is willing to do the review which I have done. Everyone is a volunteer so when things get busy in our personal lives we can miss what is going on.
RE the copyright changes, I believe there is a line to give MVN and it will make the changes for you. You will need to rebase first, see here:
https://community.openhab.org/t/compiling-oh3-thing/123511/4
Also since 3.2 Stable is released, you also need to change to 3.3 from 3.2 snapshot now in the bindings main pom.

**Comment by Sonic-Amiga** [comment]
@Skinah @fwolter I've updated version to 3.3-SNAPSHOT. Copyright messages are still intact over the whole repo, so i've kept them too. Maybe you'll be very kind to merge the code and then update everything.

**Comment by Sonic-Amiga** [comment]
Updated year to 2022

**Review by fwolter — CHANGES_REQUESTED** [review]
@Skinah thanks for your review, you made some good points!

@Sonic-Amiga There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html` after building it with `mvn clean install`.

File: bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/SerialBus.java
Comment (Sonic-Amiga) [reply bundles/org.openhab.binding.herzborg/src/main/java/org/openhab/binding/herzborg/internal/SerialBus.java:94]: Thanks, slipped back in during an update, i remember having already removed it once.

File: bundles/org.openhab.binding.herzborg/README.md
Comment (Sonic-Amiga) [reply bundles/org.openhab.binding.herzborg/README.md:43]: In fact it is a state. It can be written, it's stored in the device and it can be read back.
What it basically does is reversing meaning of 0% and 100%. By default 100% is fully closed and 0% is fully open. Setting "reverse" to true swaps that. And yes, i know that openhab actually prefers it reversed.
I agree that this is not end-user setting, it's purely technical, used only during setup. But the hardware comes with no software tools, only with protocol description, so i decided to make all the settings available, just in case. There aren't many of them, so it was easy.

File: bundles/org.openhab.binding.herzborg/README.md
Comment (Sonic-Amiga) [reply bundles/org.openhab.binding.herzborg/README.md:35]: Ok, i could and i will.

File: bundles/org.openhab.binding.herzborg/src/main/feature/feature.xml
Comment (Sonic-Amiga) [reply bundles/org.openhab.binding.herzborg/src/main/feature/feature.xml:2]: Removed

File: bundles/org.openhab.binding.herzborg/README.md
Comment (fwolter) [reply bundles/org.openhab.binding.herzborg/README.md:43]: OK

**Comment by Sonic-Amiga** [comment]
Updated, all notes fixed.
Only two analysis warnings remain: about config data classes not having NonNullByDefault, but as far as i know that's normal.

**Comment by fwolter** [comment]
The config class should have also the null annotation.

**Comment by Sonic-Amiga** [comment]
@fwolter Done, please, enjoy

**Comment by Sonic-Amiga** [comment]
@fwolter May i gently ping you ?

**Review by fwolter — APPROVED** [review]
LGTM

Now, you could add your binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website

**Comment by Sonic-Amiga** [comment]
@fwolter 
> Now, you could add your binding's logo to the openHAB website

Hello. I can take a company's .png logo from their web site and use that: http://www.herzborg.com/images/logo.png . But is it legal? I suppose a company logo is protected by license/trademark/other blah-blah-blah...

What else can i do ? Please advice me something, i am not an artist at all. I mean - zero percent. I can't draw even icons.

**Comment by fwolter** [comment]
You could kindly ask Herzborg if it's okay to add their logo with a reference to the openHAB website with all the other logos.

**Comment by Sonic-Amiga** [comment]
Okay, i will.

**Pull request merged**
